lib/livebook/update_check.ex

```diff
@@ -0,0 +1,152 @@
+defmodule Livebook.UpdateCheck do
+  @moduledoc false
+
+  # Periodically checks for available Livebook update.
+
+  use GenServer
+
+  require Logger
+
+  @name __MODULE__
+  @timeout :infinity
+
+  @hour_in_ms 60 * 60 * 1000
+  @day_in_ms 24 * @hour_in_ms
+
+  @doc false
+  def start_link(_opts) do
+    GenServer.start_link(__MODULE__, {}, name: @name)
+  end
+
+  @doc """
+  Returns the latest Livebook version if it's more recent than the
+  current one.
+  """
+  @spec new_version() :: String.t() | nil
+  def new_version() do
+    GenServer.call(@name, :get_new_version, @timeout)
+  end
+
+  @doc """
+  Returns whether the update check is enabled.
+  """
+  @spec enabled?() :: boolean()
+  def enabled?() do
+    GenServer.call(@name, :get_enabled, @timeout)
+  end
+
+  @doc """
+  Sets whether the update check is enabled.
+  """
+  @spec set_enabled(boolean()) :: :ok
+  def set_enabled(enabled) do
+    GenServer.cast(@name, {:set_enabled, enabled})
+  end
+
+  @impl true
+  def init({}) do
+    state = %{
+      enabled: Livebook.Settings.update_check_enabled?(),
+      new_version: nil,
+      timer_ref: nil
+    }
+
+    {:ok, schedule_check(state, 0)}
+  end
+
+  @impl true
+  def handle_cast({:set_enabled, enabled}, state) do
+    Livebook.Settings.set_update_check_enabled(enabled)
+    state = %{state | enabled: enabled}
+    state = state |> cancel_check() |> schedule_check(0)
+    {:noreply, state}
+  end
+
+  @impl true
+  def handle_call(:get_enabled, _from, state) do
+    {:reply, state.enabled, state}
+  end
+
+  @impl true
+  def handle_call(:get_new_version, _from, state) do
+    new_version = if state.enabled, do: state.new_version
+    {:reply, new_version, state}
+  end
+
+  @impl true
+  def handle_info(:check, state) do
+    myself = self()
+
+    Task.Supervisor.start_child(Livebook.TaskSupervisor, fn ->
+      send(myself, {:check_response, fetch_latest_version()})
+    end)
```

Comment:
I would do async_nolink and store the `task` in the state. This way we also receive the DOWN message in case of errors so we can retry in one hour. Alternatively keep as is but do a Process.monitor(...) on the returned PID.

```suggestion
    Task.Supervisor.async_nolink(Livebook.TaskSupervisor, fn ->
      fetch_latest_version()
    end)
```

Reply:
One option is even to change `fetch_latest_supervisor()` to call `exit({:shutdown, :retry})`, instead of returning `{:error, ...}`. This way you only need to handle the successful result and the down messages. We use a shutdown tuple so nothing is logged.